### PR TITLE
Add tests to cover the XmlSchema and XmlSerializerFactory scenarios.

### PR DESCRIPTION
**Pull request by huanwu**
@shmao @mconnew @zhenlan 

https://github.com/dotnet/corefx/issues/10570
https://github.com/dotnet/corefx/issues/10573

### CODE REVIEW

**Comment by huanwu** [comment]
@dotnet-bot test Innerloop CentOS7.1 Release Build and Test, Innerloop OSX Debug Build and Test, Innerloop Windows_NT Debug Build and Test, Innerloop Windows_NT Release Build and Test

File: src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 1]
```diff
@@ -1838,6 +1839,69 @@ static SimpleType[][] GetObjectwith2DArrayOfSimpleType()
         simpleType2D[1][1] = new SimpleType() { P1 = "1 1 value", P2 = 4 };
         return simpleType2D;
     }
+
+    [Fact]
+    public static void XmlSchemaTest()
+    {
+        XmlSchemas schemas = new XmlSchemas();
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs]: Nit: use var when the type of the declaration is obvious from the initializer, especially if it is an object creation. This eliminates redundancy.

File: src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 2]
```diff
@@ -1838,6 +1839,69 @@ static SimpleType[][] GetObjectwith2DArrayOfSimpleType()
...
+        Assert.Equal(originalmapping.XsdTypeName, newmapping.XsdTypeName);
+        Assert.Equal(originalmapping.XsdTypeNamespace, newmapping.XsdTypeNamespace);
+    }
+
+    [Fact]
+    public static void XmlSerializerFactoryTest()
+    {
+        string expected = "<?xml version=\"1.0\"?>\r\n<Dog xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\">\r\n  <age>5</age>\r\n  <name>Bear</name>\r\n  <breed>GermanShepherd</breed>\r\n</Dog>";
+        XmlSerializerFactory xsf = new XmlSerializerFactory(); ;
+        XmlSerializer xs = xsf.CreateSerializer(typeof(Dog));
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs]: Can you please use `SerializeAndDeserialize` method instead?

File: src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 3]
```diff
@@ -1838,6 +1839,69 @@ static SimpleType[][] GetObjectwith2DArrayOfSimpleType()
...
+        MemoryStream ms = new MemoryStream();
+        string expected = "<?xml version=\"1.0\"?>\r\n<xs:schema elementFormDefault=\"qualified\" xmlns:xs=\"http://www.w3.org/2001/XMLSchema\">\r\n  <xs:element name=\"Dog\" nillable=\"true\" type=\"Dog\" />\r\n  <xs:complexType name=\"Dog\">\r\n    <xs:complexContent mixed=\"false\">\r\n      <xs:extension base=\"Animal\">\r\n        <xs:sequence>\r\n          <xs:element minOccurs=\"1\" maxOccurs=\"1\" name=\"breed\" type=\"DogBreed\" />\r\n        </xs:sequence>\r\n      </xs:extension>\r\n    </xs:complexContent>\r\n  </xs:complexType>\r\n  <xs:complexType name=\"Animal\">\r\n    <xs:sequence>\r\n      <xs:element minOccurs=\"1\" maxOccurs=\"1\" name=\"age\" type=\"xs:int\" />\r\n      <xs:element minOccurs=\"0\" maxOccurs=\"1\" name=\"name\" type=\"xs:string\" />\r\n    </xs:sequence>\r\n  </xs:complexType>\r\n  <xs:simpleType name=\"DogBreed\">\r\n    <xs:restriction base=\"xs:string\">\r\n      <xs:enumeration value=\"GermanShepherd\" />\r\n      <xs:enumeration value=\"LabradorRetriever\" />\r\n    </xs:restriction>\r\n  </xs:simpleType>\r\n</xs:schema>";
+        while (schemaEnumerator.MoveNext())
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs]: I wondered if we want a `while` loop here. We know we have only one schema. I think we'd better Assert that in this test.

File: src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 4]
```diff
@@ -1838,6 +1839,69 @@ static SimpleType[][] GetObjectwith2DArrayOfSimpleType()
...
+        {
+            schemaEnumerator.Current.Write(ms);
+            ms.Position = 0;
+            string actualOutput = new StreamReader(ms).ReadToEnd();
+            Utils.CompareResult result = Utils.Compare(expected, actualOutput);
+            Assert.True(result.Equal, string.Format("{1}{0}Test failed for wrong output from schema: {0}Expected: {2}{0}Actual: {3}",
+                    Environment.NewLine, result.ErrorMessage, expected, actualOutput));
+        }
+
+        schemas.Compile((o, args) => ValidationEventHandle(o, args), true);
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs]: Can we replace `ValidationEventHandle` with a lambda?

File: src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs [hunk 5]
```diff
@@ -3202,3 +3202,20 @@ public class CollectionOfTypeWithNonDefaultNamcespace : List<TypeWithNonDefaultN
 
 }
 
+
+public class Animal
+{
+    public int age = 4;
+    public string name = "Rusky";
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs]: Do we need to initialize the fields here?

File: src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 6]
```diff
@@ -1838,6 +1839,54 @@ static SimpleType[][] GetObjectwith2DArrayOfSimpleType()
...
+        var ms = new MemoryStream();
+        string baseline = "<?xml version=\"1.0\"?>\r\n<xs:schema elementFormDefault=\"qualified\" xmlns:xs=\"http://www.w3.org/2001/XMLSchema\">\r\n  <xs:element name=\"Dog\" nillable=\"true\" type=\"Dog\" />\r\n  <xs:complexType name=\"Dog\">\r\n    <xs:complexContent mixed=\"false\">\r\n      <xs:extension base=\"Animal\">\r\n        <xs:sequence>\r\n          <xs:element minOccurs=\"1\" maxOccurs=\"1\" name=\"breed\" type=\"DogBreed\" />\r\n        </xs:sequence>\r\n      </xs:extension>\r\n    </xs:complexContent>\r\n  </xs:complexType>\r\n  <xs:complexType name=\"Animal\">\r\n    <xs:sequence>\r\n      <xs:element minOccurs=\"1\" maxOccurs=\"1\" name=\"age\" type=\"xs:int\" />\r\n      <xs:element minOccurs=\"0\" maxOccurs=\"1\" name=\"name\" type=\"xs:string\" />\r\n    </xs:sequence>\r\n  </xs:complexType>\r\n  <xs:simpleType name=\"DogBreed\">\r\n    <xs:restriction base=\"xs:string\">\r\n      <xs:enumeration value=\"GermanShepherd\" />\r\n      <xs:enumeration value=\"LabradorRetriever\" />\r\n    </xs:restriction>\r\n  </xs:simpleType>\r\n</xs:schema>";
+        schemaEnumerator.MoveNext();
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs]: Nit: Assert this to be true?

File: src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 7]
```diff
@@ -1838,6 +1839,54 @@ static SimpleType[][] GetObjectwith2DArrayOfSimpleType()
...
+    }
+
+    [Fact]
+    public static void XmlSerializerFactoryTest()
+    {
+        string baseline = "<?xml version=\"1.0\"?>\r\n<Dog xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\">\r\n  <age>5</age>\r\n  <name>Bear</name>\r\n  <breed>GermanShepherd</breed>\r\n</Dog>";
+        var xsf = new XmlSerializerFactory();
+        Func<XmlSerializer> serializerfunc = () => xsf.CreateSerializer(typeof(Dog));
+        var dog1 = new Dog() { name = "Bear", age = 5, breed = DogBreed.GermanShepherd };
+        var dog2 = SerializeAndDeserialize(dog1, baseline, serializerfunc, true);
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs]: Why is `skipStringCompare` set to true?

File: src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs [hunk 8]
```diff
@@ -3202,3 +3202,20 @@ public class CollectionOfTypeWithNonDefaultNamcespace : List<TypeWithNonDefaultN
 
 }
 
+
+public class Animal
+{
+    public int age;
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTypes.cs]: Nit: Capitalize public field's name.

**Comment by shmao** [comment]
LGTM.

**Pull request merged**